snapcraft_legacy/cli/_options.py

```diff
@@ -285,21 +285,50 @@ def _sanity_check_build_provider_flags(build_provider: str, **kwargs) -> None:
 
 
 def get_build_provider(**kwargs) -> str:
-    """Get build provider and determine if running as managed instance."""
-
+    """Get the appropriate provider and determine if running as a managed instance.
+
+    To determine the appropriate provider:
+    (1) use provider specified by `--provider=<provider>`
+    (2) use provider from the environment variable `SNAPCRAFT_BUILD_ENVIRONMENT`
+        (`click` loads this env var into `--provider=<provider>`)
+    (3) use provider 'lxd' if `--use-lxd` was passed
+    (4) use provider 'host' if `--destructive-mode` was passed
+    (5) use provider 'host' if snapcraft is running as a process container
+    (6) use provider specified by snap config
+    (7) else default to multipass
+
+    :return: provider name
+    :raises ValueError: if provider is not supported
+    """
+    # load provider from snap config
+    snap_config = common.get_snap_config()
+    snap_provider = snap_config.get("provider") if snap_config else None
+
```

Comment:
Bad combinations of parameters are handled elsewhere, or we should handle them here?

I mean, passing `--provider` and `--use-lxd` together should be an error. Or `--destructive-mode` with any of those.

Note that having the envvar is fine with any of those options, command line should override environmental config. Probably `click` is messing everything here.

Reply:
Those cases are checked in [`_sanity_check_build_provider_flags()`](https://github.com/snapcore/snapcraft/blob/74f1b86351d34288f16295a6257c33ca9a811e5b/snapcraft_legacy/cli/_options.py#L237), which is called at the end of this function.

The checks are not very restrictive - your example `snapcraft --provider=lxd --use-lxd` is valid.  
It does throw an error when `--destructive-mode` is used with `--provider=<provider>` or `--use-lxd`.

Reply:
Awesome, thanks!